### ISSUE TITLE
Proposal: Port EqualOpportunityClassifier and DemographicParityClassifier

### ISSUE THREAD

**Issue by MBrouns** [fairlearn/fairlearn#466]
I'm making this issue so that we can disucss how to move the `EqualOpportunityClassifier` and `DemographicParityClassifier from scikit-fairness to this project.

The methods are discussed in a notebook here: https://scikit-lego.readthedocs.io/en/latest/fairness.html

The `DemographicParityClassifier` is an implementation of the method described in M. Zafar et al. (2017), Fairness Constraints: Mechanisms for Fair Classification. The optimization itself is done using `cvxpy`. The `EqualOpportunityClassifier` is not described in a paper afaik, but is a simple modification on the constraint of the `DemographicParityClassifier`

Is this something you'd be open to having in this repo? If so, I can start working on a PR.

**Comment by riedgar-ms** [comment]
I'll defer to @MiroDudik  on the question of the algorithm itself (but I expect a positive response - we'd like a richer library!), but I'm not a huge fan of the name. We've tended to have the name of the technique be the class name, and then specify the constraint as an argument. This is obviously true for things like reductions, but we do something similar for ThresholdOptimizer.

What sub-package were you thinking of using (creating, I expect)?

**Comment by kevinrobinson** [comment]
It seems awesome to merge efforts! 👍   It'd be super helpful to understand why this approach is valuable, particularly in comparison to what is already in this repository, and other contributions being discussed.  If there's a way to show this in an example scenario or notebook, that'd be amazing.

re: fairness.html, I've shared some feedback on the fairness.html example already over in https://github.com/koaning/scikit-fairness/issues/31, and a specific critique https://github.com/fairlearn/fairlearn/issues/430#issuecomment-632690069, but to recap: if you are trying to predict the median house price in a census tract, I don't understand why you would want to enforce that those predictions are uncorrelated with information about the income of people living in that census tract.  It seems like this is what the [Demographic Parity Classifier](https://scikit-lego.readthedocs.io/en/latest/fairness.html#Demographic-Parity-Classifier) and [Equal Opportunity](https://scikit-lego.readthedocs.io/en/latest/fairness.html#Equal-opportunity) examples are doing, so I'd love to understand more about why someone would use this technique, or about how this would mitigate any real harms.

Alternately, if the idea is that the task is similar to what the Boston dataset is used for in the original [Harrison and Rubinfeld work](https://deepblue.lib.umich.edu/bitstream/handle/2027.42/22636/0000186.pdf?sequence=1&isAllowed=y), it's harder for me to understand or interpret.  The original paper is not using this dataset to do a prediction task - the point of the hedonic pricing methodology in the paper is to learn the relationships between the features they have chosen and the median housing price in that tract.  They then use the weights of that model to understand the relative influence of the "clean air" feature, and how much it contributes to the median housing price which they know as ground truth.  So in that context I'm trying to understand the rationale for applying mitigation techniques to the predictions, since their methodology doesn't involve using the dataset to make predictions.

To try to help brainstorm another idea, I noticed that the Zafar paper runs its experiments on the UCI adult income dataset, and a "bank dataset."  Perhaps it might be helpful to explore if the original paper has example tasks that can show the value of these approaches?  While the URL in the paper to the bank dataset is broken, at first glance it seems like it refers to [a scenario](https://archive.ics.uci.edu/ml/datasets/Bank+Marketing) where a bank is trying to predict the success of a telemarking phone call campaign to get people to purchase 'term deposits' like bonds.  The paper discretizes age to "between 25 and 60 or not" and then shows the technique applied to a binary classifier.  I'm happy to help brainstorm but unfortunately it's also hard for me to understand how enforcing demographic parity on predictions about the success of a telemarketing campaign contributes to mitigating real harms.  I also don't understand paper's references to "disparate impact" in the context of marketing campaigns (eg, the company is deciding how many times to call each person and each row in the "bank dataset" encodes the bank's salespeople's decisions about how many times to call that home).

To be super clear, I'd love to see this kind of merging and collaboration, and I'm just a curious outsider :)  So I'm trying to share this feedback and asking these kinds of questions in that same spirit 😄

**Comment by MBrouns** [comment]
I agree that the current names are not super great but I'm not sure I like the alternative either. Basically the current algorithm is a logistic regression with a constraint on the covariance between the predicted probabilities and the sensitive attributes, that is *a* way of imposing demographic parity, but I'm pretty sure there can be others as well. You easily run the risk of creating god objects and code that is not open/closed. I just took a quick look at the `ThresholdOptimizer` and the presence of a function named ` _threshold_optimization_demographic_parity` confirms this fear.

I think this would make most sense in a `linear_models` package, with names like `EqualizedOddsLogisticRegression` and `DemographicParityLogisticRegression`. 
 
@kevinrobinson Agreed on the point regarding the dataset in the notebook I linked. Other datasets seem much more suitable. We did also test this approach briefly with another dataset on arrests in another bit of experimentation that I did with @koaning although that was more focused on how we could do these kinds of fairness topics in a Bayesian framework. (https://www.mbrouns.com/posts/2020-02-20-prior-has-some-potential/)

**Comment by kevinrobinson** [comment]
@MBrouns thanks!  I see that post, and am guessing the dataset is the same as the one with the same number of records [included in R](https://www.rdocumentation.org/packages/carData/versions/3.0-3/topics/Arrests)?  The R docs say that it's from police records in Toronto, filtered to only include arrests made for possession of small amounts of marijuana, and it doesn't disclose when or how it was collected.

I think it's important to say that predictive policing has been the subject of extensive work by many people, so it'd be amazing to connect the assessment and mitigation techniques here with that work!  If it's helpful, I think these two papers might be the most directly relevant to what you're doing here [An algorithm for removing sensitive information: application to race-independent recidivism prediction (Johndrow and Lum 2017)](https://arxiv.org/pdf/1703.04957.pdf) and [To predict and serve? (Lum and Isaac 2016)](https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.1740-9713.2016.00960.x) is shorter and more narrative.  In particular, they discuss with the sociotechnical context around the collection of the dataset is critical in evaluating algorithmic fairness mitigation techniques because of particular forms of observation bias and feedback loops in policing patterns.

FWIW, racial bias in policing is the subject of particular attention and discussion at this moment, at least in the US.  So potential users and audiences for fairlearn may be even more attuned to evaluating any algorithmic approaches that aim to reduce harm in predictive policing systems in a wider sociotechnical context.

**Comment by romanlutz** [comment]
> You easily run the risk of creating god objects and code that is not open/closed. I just took a quick look at the ThresholdOptimizer and the presence of a function named _threshold_optimization_demographic_parity confirms this fear.

Just to quickly address this fear: there's a PR by @MiroDudik out right now that generalizes this so that these things aren't specific under the hood. I don't know to what extent this is possible but can you imagine extending it to other constraints (say, true positive parity) in the future? I have yet to read the paper, so I can't say right now. If this is something we're interested in, then perhaps the "god object" as you call it isn't the worst idea. The APIs in the repo so far are written in a way where you specify the constraint as an argument in the constructor, so these would be the first ones that have it in the name.
I'm curious what you think about this.

---

I think @riedgar-ms referred to the general name when he wondered whether there's perhaps something better. For example, you could call any classifier that optimizes for Demographic Parity `DemographicParityClassifier`, but it really should be descriptive of the technique that's used.

> I think this would make most sense in a linear_models package, with names like EqualizedOddsLogisticRegression and DemographicParityLogisticRegression.

This sounds much better to me because it describes the technique better.

**Comment by riedgar-ms** [comment]
Yes, I was wondering if there was something equivalent to the `Moment` objects for this algorithm, and if that pattern could be used. It would be nice to have something like:
`est = ConstrainedLogisticRegression( constraint=DemographicParity(), ....)`
although I suspect that `DemographicParity` would not be from our `Moment` hierarchy (since we don't want those becoming god-objects either).

**Comment by romanlutz** [comment]
@riedgar-ms those are reductions-method specific. `ThresholdOptimizer` will have string arguments to specify the constraints, at least post-Miro's-PR.

**Comment by riedgar-ms** [comment]
> @riedgar-ms those are reductions-method specific. `ThresholdOptimizer` will have string arguments to specify the constraints, at least post-Miro's-PR.

Indeed - but if there is useful algorithmic work which can be done inside the 'constraint' objects, the same general pattern can be followed, surely? What I don't know is whether there is sufficient commonality to be worthwhile.

**Comment by MBrouns** [comment]
@romanlutz @MiroDudik Did you have a look at the implementation in scikit-lego? Is there (apart from naming) any concerns that should be addressed before porting?

**Comment by romanlutz** [comment]
I have to read it in depth, but I don't see a concern since there's a paper backing it up. I'll definitely prioritize this once it's a PR, so please go ahead.

Just to make sure, though: your implementation is different from the paper authors' right? I stumbled upon this repo while looking for the paper https://github.com/mbilalzafar/fair-classification and they have this licensed under GPL which is very restrictive, so I just want to make sure. It doesn't strike me as similar, but I haven't had the time for a closer look yet. I'm curious if you have any idea how your implementation is different. Perhaps they are willing to give us an exception from their license or something like that if theirs has any properties that are useful.

**Comment by MBrouns** [comment]
The implementation is indeed different, I have just taken a look through their codebase and it seems like they also support a linear SVM and they also seem to support different constraints, although I'm not really sure yet what those are.

They do use the same solver (cvxpy). 

The implementation of the `EqualOpportunityLogisticRegression` is not backed by a paper. I'll see if I can submit a PR this week, but it will basically be a 1-to-1 port of what's available in scikit-lego. That's why I wanted to check whether there's concerns beforehand.

**Comment by MiroDudik** [comment]
I looked at the proposed algorithm and I agree that this would be a nice addition (and our first inprocessing algorithm). It solves similar problems as the reductions, but the optimization problems it creates might be simpler to solve in practice.

As you say, the Demographic Parity follows
* [Fairness Constraints: Mechanisms for Fair Classification](https://arxiv.org/pdf/1507.05259.pdf)

Your Equal Opportunity implementation actually follows another approach by Zafar et al:
* [Fairness Beyond Disparate Treatment & Disparate Impact: Learning Classification without Disparate Mistreatment](https://arxiv.org/pdf/1610.08452.pdf)
(see their Eq. 13, where _g<sub>&#952;</sub>_ is set according to Eq. 12, to reflect Eq. 5; note that they encode negative examples as y=-1)

We should think more about the right API, but the algorithm is a go from my perspective.

My current thoughts about the API... I am thinking about something like:
```
from fairlearn.inprocessing import ConstrainedLogisticRegression

est = ConstrainedLogisticRegression('demographic_parity', ...)
est = ConstrainedLogisticRegression('true_positive_rate_parity', ...)
```
But I think that we may want to consider changing the name, to e.g. something like:
* `'demographic_parity'` -> `'bounded_covariance'` with a keyword argument `'upper_bound'`

Do we want to continue discussing API on this issue or would it be better to have a call?

**Comment by koaning** [comment]
Why not 

```
from fairlearn.linear_model import ConstrainedLogisticRegression
```

**Comment by MBrouns** [comment]
> 'demographic_parity' -> 'bounded_covariance' with a keyword argument 'upper_bound'

Both the equal opportunity and demographic parity have bounded covariance constraints though. How would we distinguish between them?

**Comment by MiroDudik** [comment]
**Re. modules**: I can see arguments either way:
* linear_model PROS:
  * the same name as in sklearn, so benefit to familiarity
* linear_model CONS:
  * it is possible to mitigate unfairness in logistic regression via reductions or postprocessing or preprocessing, so I don't think that we necessarily what the users to go instinctively to `fairlearn.linear_model` for these needs (?)
  * I don't think that we should try to have a mitigated variant for all estimators in `sklearn.linear_model` 
* inprocessing PROS:
  * for better or worse, the `inprocessing` is now a terminology accepted in parts of the fair ML community, e.g., https://arxiv.org/pdf/1908.09635.pdf and https://aif360.readthedocs.io/en/latest/modules/algorithms.html#module-aif360.algorithms.inprocessing
* inprocessing CONS:
  * I don't necessarily love the name... also, this might become a bit too much of a catch-all module

**Re. function API**: Good points re. covariance, how about the following format:
```
est = ConstrainedLogisticRegression('demographic_parity', covariance_bound=0.1, ...)
est = ConstrainedLogisticRegression('true_positive_rate_parity', covariance_bound=0.1, ...)
```
or maybe:
```
est = ConstrainedLogisticRegression(constraints='demographic_parity', covariance_bound=0.1, ...)
est = ConstrainedLogisticRegression(constraints='true_positive_rate_parity', covariance_bound=0.1, ...)
```

I think that this would play nicely with how we do things elsewhere, e.g.,
```
est = ExponentiatedGradient(base_estimator, DemographicParity(difference_bound=0.1), ...)
est = ExponentiatedGradient(base_estimator, DemographicParity(ratio_bound=0.8), ...)
est = ThresholdOptimizer(estimator=base_estimator, constraints='demographic_parity', ...)
```

**Comment by MiroDudik** [comment]
A quick update: I have taken an informal poll among a few fair ML / data mining researchers, and folks raised several additional issues with in-processing (for example, it sounds like data transformation rather than training and so it is confusing--also, it is not as widely adopted as I thought), so I no longer think that we should consider it as a module name. So let's stick to **fairlearn.linear_models**.

We just need to settle the calling API. Let's first figure out how to handle specification of constraints. Let me repeat my two suggestions from above and add a third one:

```
class ConstrainedLogisticRegression(BaseEstimator, LinearClassifierMixin):
    def __init__(self, constraints='demographic_parity', *, covariance_bound=0.0, ...)

class ConstrainedLogisticRegression(BaseEstimator, LinearClassifierMixin):
    def __init__(self, *, constraints='demographic_parity', covariance_bound=0.0, ...)

class LogisticRegression(BaseEstimator, LinearClassifierMixin):
    def __init__(self, *, constraints=None, covariance_bound=0.0, ...)
```
The idea is that `constraints=None` would just revert to the unconstrained behavior. I think that your current implementation has the default corresponding to `covariance_bound=np.inf`--I think it's a bit counter-intuitive, but open to discussion.

Thoughts?

**Comment by MiroDudik** [comment]
@MBrouns : Ping. Let us know if you want to resume working on this, or prefer to drop it, or maybe pitch to the community to take over?

**Comment by romanlutz** [comment]
@MBrouns any thoughts? :-)

**Comment by MBrouns** [comment]
Apologies for the radio silence. It's been busy and I've been avoiding programming in my spare time a bit to focus on other things. 

I'm still happy to pick this up but I can't commit to a deadline. If there's someone lined up who wants to tackle this I'm happy for them to give it a go. If not, I'll try to make a start near the end of the month

**Comment by koaning** [comment]
I also don't have time, but probably a topic that requires discussing upfront: our implementation of these tools requires a dependency from cvxpy. @MiroDudik is this a dependency you're willing to accept?

**Comment by romanlutz** [comment]
There's no particular urgency AFAIK. @hildeweerts and I have just been going through all issues to add more detail where possible or mark as "help wanted" where appropriate, since we're going to participate in developer sprints at PyCon and SciPy this year. 

Thanks for pointing out the `cvxpy` dependency @koaning ! Out of interest I've installed `cvxpy` locally and it brought in a few new ones I didn't have: `ecos`, `scs`, `osqp`, and `qdldl`. This is the `pipdeptree` output in case you're interested:
```
cvxpy==1.1.11
  - ecos [required: >=2, installed: 2.0.7.post1]
    - numpy [required: >=1.6, installed: 1.19.0]
    - scipy [required: >=0.9, installed: 1.4.1]
      - numpy [required: >=1.13.3, installed: 1.19.0]
  - numpy [required: >=1.15, installed: 1.19.0]
  - osqp [required: >=0.4.1, installed: 0.6.2.post0]
    - numpy [required: >=1.7, installed: 1.19.0]
    - qdldl [required: Any, installed: 0.1.5.post0]
      - numpy [required: >=1.7, installed: 1.19.0]
      - scipy [required: >=0.13.2, installed: 1.4.1]
        - numpy [required: >=1.13.3, installed: 1.19.0]
    - scipy [required: >=0.13.2, installed: 1.4.1]
      - numpy [required: >=1.13.3, installed: 1.19.0]
  - scipy [required: >=1.1.0, installed: 1.4.1]
    - numpy [required: >=1.13.3, installed: 1.19.0]
  - scs [required: >=1.1.6, installed: 2.1.2]
    - numpy [required: >=1.7, installed: 1.19.0]
    - scipy [required: >=0.13.2, installed: 1.4.1]
      - numpy [required: >=1.13.3, installed: 1.19.0]
```
... definitely makes me lean towards soft dependency (which is what we do for things like `matplotlib` right now).
@adrinjalali typically has thoughts on such matters, too :-)

**Comment by adrinjalali** [comment]
yeah I've had issues with `cvxpy` before because of its dependencies. It's the kinda package that I would be really worried about for having a hard dependency on.

I would suggest having it as another soft dependency, and having on `pip install fairlearn[complete]` for instance. But we should also start having a very clear table, with methods/modules on one side, and dependencies on the other side, for people to have a clear view about them.

**Comment by MiroDudik** [comment]
I like @adrinjalali 's suggestion to have `cvxpy` as a soft dependency. I like the suggestion re. doc table with soft dependencies. I'm agnostic if we need `complete` install variant--but it seems nice from the user perspective.
